Enable scoped storage enforcement on Android 11+.

hrydgard opened the pull request:
Targets Android SDK 30, so scoped storage gets turned on for real on Android 11+ devices, see #11997.

~~This has a number of UX issues and bugs we need to work through, but at least games start up. ~~ Upgrades are handled smoothly by keeping existing storage access until you uninstall. After a reinstall, you'll need to re-select your old PSP directory manually in settings :(

We need to merge this soon, since after August we can no longer release new PPSSPP updates on Google Play without doing this. Most people should not be badly affected once this is in, just a new procedure to learn to give PPSSPP permissions to the folders you want to access on first startup.

We absolutely need to merge this AND release a new PPSSPP version with it before whenever Android 12 releases in September, October or November, and the old ways will no longer work for new installs.

This will take care of #11997 once done. Again, do not merge yet!

## Sub-issues discovered and fixed:

- /PSP as content URI:
  - [x] Save states not saving
  - [x] Save state thumbnails not showing
  - [x] In-game saves not working (game tested: Wipeout Pulse)
- Other
  - [x] File listing performance
  - [x] The code for determining free storage space is broken
  - [x] File move not implemented
  - [x] File append not working, probably some other edge cases
  - [x] Add support for returning error codes
  - [ ] Actually catch exceptions and provide proper error codes for things like disk full
  - [x] Make things work as expected when the memstick folder is named PSP 
- Planned features
  - [x] On first run, ask the user where they want their PSP folder. 
    - [x] Get it working
    - [ ] Make it pretty and make sure the user understands the choice
  - [x] Ability to auto-migrate your data when changing your PSP folder
  - [x] Threaded file move with progress bar
  - [ ] Set correct MIME types? Worth the trouble?

## Setup screen
If we find no folder configured and scoped storage enabled (first boot, or reinstall on Android11+) we will pop up a screen that gives the user two options, along with some advice:

* Choose or create a folder (browse button)
  - Choose if you had an old PPSSPP version and thus an old PSP folder
  - Allows sharing data between PPSSPP Regular and PPSSPP Gold

* Default PSP folder location
  - Will be removed if you uninstall PPSSPP

Not sure what else to say. There seem to be little reason to use the default folder location really, now that performance is pretty good in both cases...

hrydgard committed:
Enable scoped storage enforcement on Android 11+.

This has a number of UX issues and bugs we need to work through, but at least
games are playable, things mostly work. Upgrades are handled smoothly by
keeping existing storage access until you uninstall. After a reinstall, you'll
need to re-select your old PSP directory manually in settings :(

hrydgard commented:
Making progress here, in-game saves are working. 

Getting quite terrible stutter around save dialogs etc, hopefully the planned optimization will be good enough...

hrydgard commented:
Now that most of this is merged, cut it back down to just the enabling commit.

hrydgard commented:
With those two commits, performance is now pretty close to acceptable, actually.

hrydgard commented:
![image](https://user-images.githubusercontent.com/130929/126811168-82cfdfaf-3162-4151-802e-e2f8cc842c24.png)

First cut at a startup experience. Ugh. Gonna need some design work..

hrydgard committed:
Get the new setup flow going, with a confirmation dialog etc.

Moving the folder while already installed still crashes.

hrydgard committed:
Basic flow of setting up and moving the PSP directory works now.

hrydgard commented:
It was suggested by unknown that maybe it's better to directly default to the internal directory and add a warning somewhere, allowing the user to choose a permanent location. 

I couldn't quite figure out how to design that though, so for the initial merge (possibly not the official release) I'm going with a popup screen offering the choice, as above. For moves after the install, there'll be a second screen where you confirm the details of the move, and that will have a progress bar etc if you are moving data from a previous location.

Such a popup screen is not really much worse than the old obnoxious ask-for-permissions screen/button, I figure, anyway.

Most things seem to be working okay now, so I'm actually gonna look into merging this fairly soon so it gets some exercise. Of course, this can easily screw up a user's data, so best be careful and test properly before merging...

hrydgard committed:
Work on the new experience

hrydgard commented:
Shrunk this again by pulling out a bunch of stuff into #14642 , should help any bisect.

Now the entire flow seems to be working okay, including moving the directory around after the fact. You have the option to move the files or not, but can't control overwrite in case of collisions yet - not sure how desirable that is, but I guess we could warn if you're about to overwrite any newer files...

hrydgard commented:
This is essentially feature complete, but needs a bunch of polish, see the missing checkboxes above. Even so, I think this should actually go in soon almost as-is so it gets as much testing as possible. It really isn't much slower anymore and seems to work quite alright. Of course, before I press merge I'm gonna subject this to a lot more manual testing, too.

Anyway, marking as ready for review, for the so inclined.

unknownbrackets commented on android/AndroidManifest.xml:
```diff
@@ -45,7 +45,9 @@
         android:banner="@drawable/tv_banner"
-        android:requestLegacyExternalStorage="true">
+        android:requestLegacyExternalStorage="true"
+		android:preserveLegacyExternalStorage="true"
+		>
```
nit: Indentation different.

-[Unknown]

unknownbrackets commented on Common/File/DiskFree.cpp:
```diff
@@ -43,7 +43,7 @@ bool free_disk_space(const Path &path, uint64_t &space) {
 #ifndef __ANDROID__
 		if (diskstat.f_flag & ST_RDONLY) {
-			space = 0;
+			space = -1;
 			return true;
```
It returns true though, isn't it better to interpret that as 0 bytes free?  We'll cast this to unsigned and try to write.

Otherwise it should at least return false?  But this isn't really an error.  The free space for a game to create new save data is 0 bytes.  It's a known value, so I'd consider it successful.

-[Unknown]

unknownbrackets commented on UI/NativeApp.cpp:
```diff
@@ -339,12 +342,13 @@ static void PostLoadConfig() {
-	// On Android, create a PSP directory tree in the external_dir,
-	// to hopefully reduce confusion a bit.
+bool CreateDirectoriesAndroid() {
+	// TODO: Really not sure why this code is Android-exclusive, except the ".nomedia" part.
```
You added this for permissions iirc because it needed to happen after permissions were granted.

-[Unknown]

hrydgard replied:
Right..

hrydgard committed:
Delete some left-behind lines

unknownbrackets commented on UI/MemStickScreen.h:
```diff
@@ -30,19 +30,52 @@
+	~MemStickScreen() {}
 
 	std::string tag() const override { return "game"; }
+	void CreateViews() override;
```
Hm, what's the reason for this to be public?  Only RecreateViews() should get called right?

-[Unknown]

hrydgard replied:
No reason, fixing.

unknownbrackets commented on UI/MemStickScreen.cpp:
```diff
@@ -25,23 +25,83 @@
+		}
+		if (!File::WriteDataToFile(true, "1", 1, testWriteFile)) {
+			// settingInfo_->Show(sy->T("ChangingMemstickPathInvalid", "That path couldn't be used to save Memory Stick files."), nullptr);
+			// TODO: Display an error!
```
Good todos are good.

-[Unknown]

unknownbrackets commented on UI/MemStickScreen.cpp:
```diff
@@ -25,23 +25,83 @@
+	} else {
+		// TODO: Do the same but with scoped storage? Not really necessary, right? If it came from a browse
+		// for folder, we can assume it exists and is writable, barring wacky race conditions like the user
+		// being connected by USB and deleting it.
```
I wonder in what exciting ways we would crash if they did that, then started PPSSPP up again later without the USB device connected...

-[Unknown]

hrydgard replied:
We would fail the check in NativeApp.cpp "Couldn't read directory '%s' specified by memstick_dir.txt", and it would ask the user for a new directory.

unknownbrackets commented on UI/MemStickScreen.cpp:
```diff
@@ -25,23 +25,83 @@
+
+static std::string FormatSpaceString(int64_t space) {
+	if (space >= 0) {
+		// TODO: Smarter display (MB, GB as appropriate). Don't we have one of these somewhere?
```
Probably should be in Common:
https://github.com/hrydgard/ppsspp/blob/0c40e918c92b897f745abee0d09cf033a1572337/Windows/W32Util/Misc.cpp#L56

But also:
https://github.com/hrydgard/ppsspp/blob/87dcf60ed037d5f2168d5f0033dda002b098e177/Core/Dialog/SavedataParam.cpp#L1050

-[Unknown]

hrydgard replied:
Right.. I'll just move NiceSizeFormat out to Common for now, could merge them later but they're slightly different (rounding, decimal...)

unknownbrackets commented on UI/MemStickScreen.cpp:
```diff
@@ -51,45 +111,50 @@ void MemStickScreen::CreateViews() {
 
-	root_->Add(settingInfo_);
+	leftColumn->Add(new Choice(iz->T("Use App Private Directory")))->OnClick.Handle(this, &MemStickScreen::OnUseInternalStorage);
+	leftColumn->Add(new TextView(iz->T("InternalStorageDesc", "* Warning! Data will be deleted if you uninstall PPSSPP!\n* Data cannot be shared with PPSSPP Gold\n* USB access through Android/data/org.ppsspp.ppsspp/files"), ALIGN_LEFT, false));
```
How well does the ini parser handle embedded newlines?

-[Unknown]

hrydgard replied:
It does handle it, if I remember correctly. but better check....

hrydgard replied:
yeah, at least on the right side, heh.

hrydgard committed:
Android 28 was the wrong cutoff for oldschool permissions, 29 is correct.

hrydgard committed:
Address feedback

hrydgard committed:
UI message cleanup

hrydgard commented:
Just tested this branch on an Android 4.0 device out of curiosity, works just fine. So we can keep our extreme compatibility range no problem (not surprised though but good to check).

hrydgard committed:
Indentation fix

unknownbrackets commented on Common/File/DiskFree.cpp:
```diff
@@ -41,9 +41,11 @@ bool free_disk_space(const Path &path, int64_t &space) {
 	int res = statvfs(path.c_str(), &diskstat);
 
 	if (res == 0) {
+		// Not sure why we're excluding Android here anyway...
```
https://github.com/hrydgard/ppsspp/commit/f687613d6875a0cb95eb35ea671e71bcc6a36aa6

IIRC, the flag field wasn't there back then, at least...

-[Unknown]

hrydgard committed:
When changing the PSP folder, move the files in a nice background process, showing progress.

Not everything is properly refreshed after the move though - for example
the main screen needs a refresh.

hrydgard committed:
Handle an empty path in the game browser.

hrydgard commented:
This is working well now, and I will merge it later this week after some more testing.

hrydgard commented:
Failed test results so far:

* Galaxy S8 (Android 9, API 28): Misdetecting scoped storage, broken folder browser and wacky content URIs. Should have been using the old ways.

hrydgard committed:
Correct the cutoff for scoped storage. Android 9,10 (API 28,29) shouldn't use it.

hrydgard committed:
Minor improvements on Android 9/10. Let the user reset to root of storage.

This includes handling weird "raw:/" content URIs. These I haven't seen
on newer versions.

hrydgard committed:
A bunch of small scoped storage UX improvements.

hrydgard commented:
Tested the flow again on some devices and made a few minor UX improvements.

This will be merged tomorrow.

hrydgard committed:
Buildfix, bugfix

hrydgard commented:
YOLO

iota97 commented on UI/MemStickScreen.cpp:
```diff
@@ -51,49 +112,91 @@ void MemStickScreen::CreateViews() {
-		freeSpaceAtMemStick = Android_GetFreeSpaceByFilePath(path.ToString());
+	if (!System_GetPropertyBool(SYSPROP_ANDROID_SCOPED_STORAGE)) {
+		leftColumn->Add(new Choice(iz->T("Use PSP folder at root of storage")))->OnClick.Handle(this, &MemStickScreen::OnUseStorageRoot);
+		leftColumn->Add(new TextView(iz->T("ChooseFolderDesc", "* Data will stay even if you uninstall PPSSPP.\n* Data can be shared with PPSSPP Gold\n* Easy USB access"), ALIGN_LEFT, false));
```
`Data can be shared with PPSSPP Gold` doesn't work really well for Gold version itself, maybe something like: "Data can be shared between normal and Gold version of PPSSPP".

The path `Android/data/org.ppsspp.ppsspp/files` displayed below should be different for the gold version too to avoid confusion.

hrydgard replied:
Oh yeah, I noticed this a few days ago and then completely forgot to fix it. Thanks!
